components/RiskFeed/Events/index.tsx

```diff
@@ -43,65 +49,73 @@ export default function Events({ title, empty, data, calls, isLoading }: Props)
           <>
             <Skeleton variant="rectangular" height={48} sx={{ borderRadius: 2 }} />
           </>
-        ) : data.count === 0 && calls.length === 0 ? (
+        ) : data.length === 0 ? (
           <Typography textAlign="center" color="grey.400">
             {empty}
           </Typography>
         ) : (
-          merge(data.results, calls).flatMap((e) =>
-            e.type === 'transaction'
-              ? [<Event key={e.data.id} tx={e.data} />]
-              : e.type === 'call'
-              ? [<EventCall key={e.data.id} call={e.data} />]
-              : [],
+          data.flatMap((e) =>
+            !e.execution && !e.schedule && e.events && e.events.length > 0 ? (
+              <EventCall calls={e.events} />
+            ) : e.execution || e.schedule ? (
+              <Event entry={e} />
+            ) : null,
           )
         )}
       </Box>
     </Box>
   );
 }
 
-function merge(
-  safe: SafeResponse['results'],
-  calls: Call[],
-): ({ type: 'transaction'; data: Transaction } | { type: 'call'; data: Call })[] {
-  const transactions = safe.flatMap((tx) => (tx.type === 'TRANSACTION' ? [tx.transaction] : []));
-  return [
-    ...transactions.map((tx) => ({ type: 'transaction', data: tx }) as const),
-    ...calls.map((c) => ({ type: 'call', data: c }) as const),
-  ].sort((x, y) => {
-    if (x.type === 'call' && y.type === 'call') {
-      if (x.data.executedAt && y.data.executedAt) {
-        return y.data.executedAt - x.data.executedAt;
-      }
-      return y.data.scheduledAt - x.data.scheduledAt;
-    }
-    if (x.type === 'call' && y.type === 'transaction') {
-      if (x.data.executedAt) {
-        return y.data.timestamp / 1000 - x.data.executedAt;
-      }
-      return y.data.timestamp / 1000 - x.data.scheduledAt;
-    }
-    if (x.type === 'transaction' && y.type === 'call') {
-      if (y.data.executedAt) {
-        return y.data.executedAt - x.data.timestamp / 1000;
-      }
-      return y.data.scheduledAt - x.data.timestamp / 1000;
+export function group(safe: SafeResponse, calls: Call[]): Entry[] {
+  const grouped = calls.reduce(
+    (store, call) => {
+      const timestamp = call.executedAt ?? call.scheduledAt;
+      store[timestamp] = store[timestamp] ? [...store[timestamp], call] : [call];
+      return store;
+    },
+    {} as Record<number, Call[]>,
+  );
+
+  const transactions = safe.results.flatMap((tx) => (tx.type === 'TRANSACTION' ? [tx.transaction] : []));
+  const leftovers = new Set(transactions);
+
+  const entries: Entry[] = [];
+
+  for (const events of Object.values(grouped)) {
+    if (events.length === 0) {
+      continue;
     }
-    if (x.type === 'transaction' && y.type === 'transaction') {
-      return y.data.timestamp - x.data.timestamp;
+
+    const call = events[0];
+    const entry: Entry = { events, timestamp: call.executedAt ?? call.scheduledAt };
+
+    entry.schedule = transactions.find((tx) => tx.timestamp / 1000 === call.scheduledAt);
```

Comment:
isn't there a need to round the result of the division?

Reply:
No, Safe uses `block.timestamp * 1000` for timestamp. All timestamps end with `000`.